modules/mod-deep-learning/ModelCard.h

```diff
@@ -0,0 +1,212 @@
+   friend class DeepModelManager;
+   
+   // throws InvalidModelCardDocument if the given json is not valid. 
+   void DeserializeFromFile(const std::string& path, const Doc& schema);
+   void SerializeToFile(const std::string& path) const;
+
+   template < typename Writer >
+   void Serialize(Writer &writer) const;
+   void Deserialize(const Doc& doc, const Doc& schema);
+
+   void Validate(const Doc& doc, const Doc& schema);
+
+   bool IsLocal() { return m_is_local; }
+   void SetLocal(bool local) { m_is_local = local; }
+
+   std::string GetLocalPath() const { return m_local_path; }
+   void SetLocalPath(const std::string& path) { m_local_path = path; }
+
+public:
+
+   std::string name() const { return m_name; }
+   void name(const std::string &name) { m_name = name; }
+
+   std::string author() const { return m_author; }
+   void author(const std::string &author) { m_author = author; }
+
+   std::string long_description() const { return m_long_description; }
+   void long_description(const std::string &long_description) { m_long_description = long_description; }
+
+   std::string short_description() const { return m_short_description; }
+   void short_description(const std::string &short_description) { m_short_description = short_description; }
+
+   int sample_rate() const { return m_sample_rate; }
+   void sample_rate(int rate) { m_sample_rate = rate; }
+
+   bool multichannel() const { return m_multichannel; }
+   void multichannel(bool multichannel) { m_multichannel = multichannel; }
+
+   std::string effect_type() const { return m_effect_type; }
+   void effect_type(const std::string &type) { m_effect_type = type; }
+
+   const std::vector<std::string> domain_tags() const { return m_domain_tags; }
+   void domain_tags(const std::vector<std::string> tags) { m_domain_tags = tags; }
+
+   const std::vector<std::string> tags() const { return m_tags; }
+   void tags(const std::vector<std::string> tags) { m_tags = tags; }
+
+   const std::vector<std::string> labels() const { return m_labels; }
+   void labels(std::vector<std::string> labels) { m_labels = labels; } // use std::move to not copy
+
+   size_t model_size() const { return m_model_size; }
+   void model_size(size_t model_size) { m_model_size = model_size; }
+
+private:
+   std::string m_name {""};
+   std::string m_author {""};
+   std::string m_long_description {""};
+   std::string m_short_description {""};
+   int m_sample_rate {0};
+   bool m_multichannel  {false};
+   std::string m_effect_type {""};
+   std::vector<std::string> m_domain_tags;
+   std::vector<std::string> m_tags;
+   std::vector<std::string> m_labels;
+   size_t m_model_size {0};
+
+   bool m_is_local {false};
+   std::string m_local_path {""};
+
+public:
+
+   // TODO: add versioning?
+   bool operator==(const ModelCard& that) const;
+   bool operator!=(const ModelCard& that) const;
+};
+
+using ModelCardHolder = std::shared_ptr<ModelCard>;
+using ModelCardFilter = std::function<bool(ModelCardHolder card)>;
```

Comment:
I see from uses elsewhere that these predicates assume a non-null pointer, and the fact that the pointer is a smart one is irrelevant.  Also the card isn't mutated.

Say all of this with types.  Use a reference to const, not a pointer to modifiable.  Make the type `std::function<bool(const ModelCardHolder&)>`

This complicates the implementation of Filter() below, but makes the interface in this header file simpler to understand.  I know more about what the predicate is and isn't allowed to do.